src/net/neighbors/walk.rs

```diff
@@ -0,0 +1,2003 @@
+                    // the DB, since it's cur_neighbor)
+                    test_debug!(
+                        "{:?}: received HandshakeAccept from {:?}",
+                        network.get_local_peer(),
+                        &message.to_neighbor_key(&data.handshake.addrbytes, data.handshake.port)
+                    );
+                    self.handle_handshake_accept_from_ping(network, &message.preamble, data, None)?;
+                }
+                StacksMessageType::StackerDBHandshakeAccept(ref data, ref db_data) => {
+                    // this peer is still alive -- will not replace it
+                    // save knowledge to the peer DB (NOTE: the neighbor should already be in
+                    // the DB, since it's cur_neighbor)
+                    test_debug!(
+                        "{:?}: received StackerDBHandshakeAccept from {:?}: {:?}, {:?}",
+                        network.get_local_peer(),
+                        &message.to_neighbor_key(&data.handshake.addrbytes, data.handshake.port),
+                        data,
+                        db_data
+                    );
+                    self.handle_handshake_accept_from_ping(
+                        network,
+                        &message.preamble,
+                        data,
+                        Some(db_data),
+                    )?;
+                }
+                StacksMessageType::Nack(ref data) => {
+                    // evict
+                    debug!(
+                        "{:?}: Neighbor {:?} NACK'ed Handshake with code {:?}; will evict",
+                        network.get_local_peer(),
+                        nkey,
+                        data.error_code
+                    );
+                    self.comms.add_broken(network, &nkey);
+                }
+                _ => {
+                    // unexpected reply -- this peer is misbehaving and should be replaced
+                    debug!("{:?}: Neighbor {:?} replied an out-of-sequence message (type {}); will replace", network.get_local_peer(), &nkey, message.get_message_name());
+                    self.comms.add_broken(network, &nkey);
+                }
+            }
+        }
+
+        if self.comms.count_inflight() > 0 {
+            // still pending
+            return Ok(None);
+        }
+
+        // done getting pings.  do our replacements
+        self.neighbor_db.replace_neighbors(
+            network,
+            &self.neighbor_replacements,
+            &mut self.result,
+        )?;
+
+        // advance state!
+        self.set_state(network.get_local_peer(), NeighborWalkState::Finished)?;
+
+        // calculate the walk result
+        if let Some(next_neighbor) = self.next_neighbor.take() {
+            // stepped! return the result
+            return Ok(Some(self.reset(
+                network.get_local_peer(),
+                next_neighbor,
+                false,
+            )));
+        }
```

Comment:
I think this is a bug.  What ought to happen is `self.walk_outbound` should remain stable over the course of the walk's lifetime, since this value indicates whether or not the first-ever neighbor stepped to was an inbound neighbor (which may or may not be NAT'ed) or an outbound neighbor (which is not NAT'ed from our vantage point).  We should be passing `self.next_walk_outbound` instead of `false`, and `self.next_walk_outbound` should be set in `step()` if we couldn't take any action in MHRWDA.